File: Java/Medium/Unique-Paths-II.java
class Solution {
    public int uniquePathsWithObstacles(int[][] obstacleGrid) {
        int m =obstacleGrid.length;
        int n = obstacleGrid[0].length;
        int[][] table = new int[m][n];
        if (obstacleGrid[0][0] == 1 || obstacleGrid[m-1][n-1] == 1) {
            return 0;
        }
        for (int r = 0; r<m; r++) {
            for (int c = 0; c<n; c++) {
                if (r == 0 && c == 0) {
                    table[r][c]=1;
                
                } else if (obstacleGrid[r][c]==1) {
                    table[r][c]=0;
                } else if (r == 0) {
                    if (table[r][c-1]==0) {
                        table[r][c]=0;
                    } else {
                        table[r][c] = 1;
                    }  
                } else if (c == 0) {
                    if (table[r-1][c]==0) {
                        table[r][c]=0;
                    } else {
                        table[r][c] = 1;
                    }  
                }  
                else {
                    table[r][c] = table[r-1][c]+table[r][c-1];
                }
            }
        }
        return table[m-1][n-1];
    }
}

File: Python/Medium/Unique-Paths-II.py
class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        n = len(obstacleGrid[0])
        m = len(obstacleGrid)
        table = [[0 for _ in range(n)] for _ in range(m)]
        for r in range(m):
            for c in range(n):
                if (r==0 and c== 0 and obstacleGrid[r][c]!=1):
                    table[r][c] = 1
                elif (obstacleGrid[r][c]==1):
                    table[r][c]=0
                elif (r == 0 and c>0):
                    table[r][c] = table[r][c-1]
                elif (c == 0 and r>0):
                    table[r][c] = table[r-1][c]
                else: 
                    table[r][c]=table[r-1][c]+table[r][c-1]
        return table[m-1][n-1]
